bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -55,10 +93,117 @@ impl From<IdParseError> for AuthenticationError {
     }
 }
 
+impl From<OidcError> for AuthenticationError {
+    fn from(e: OidcError) -> AuthenticationError {
+        AuthenticationError::OidcError { message: e.to_string() }
+    }
+}
+
+/// The configuration to use when authenticating with OIDC.
+#[derive(uniffi::Record)]
+pub struct OidcConfiguration {
+    /// The name of the client that will be shown during OIDC authentication.
+    pub client_name: String,
+    /// The redirect URI that will be used when OIDC authentication is
+    /// successful.
+    pub redirect_uri: String,
+    /// A URI that contains information about the client.
+    pub client_uri: String,
+    /// A URI that contains the client's terms of service.
+    pub tos_uri: String,
+    /// A URI that contains the client's privacy policy.
+    pub policy_uri: String,
+
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    pub static_registrations: HashMap<String, String>,
+}
+
+/// The data needed to restore an OpenID Connect session.
+#[derive(Debug, Serialize, Deserialize)]
+struct OidcRegistrations {
+    /// The path of the file where the registrations are stored.
+    file_path: PathBuf,
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    static_registrations: HashMap<String, String>,
+}
+
+/// Manages the storage of OIDC registrations.
+impl OidcRegistrations {
+    fn new(
+        base_path: &str,
+        static_registrations: HashMap<String, String>,
+    ) -> Result<Self, AuthenticationError> {
+        let oidc_directory = PathBuf::from(base_path).join("oidc");
+        fs::create_dir_all(&oidc_directory).map_err(|_| AuthenticationError::InvalidBasePath)?;
+
+        Ok(OidcRegistrations {
+            file_path: oidc_directory.join("registrations.json"),
+            static_registrations,
+        })
+    }
+
+    /// Returns all of the registrations this client has made as a HashMap of
+    /// issuer URL (as a string) to client ID (as a string).
+    fn dynamic_registrations(&self) -> HashMap<String, String> {
+        let Some(file) = File::open(&self.file_path).ok() else {
+            return HashMap::new();
+        };
+        let reader = BufReader::new(file);
```

Comment:
We should log the error.

```suggestion
        let reader = match File::open(&self.file_path) else {
            Ok(file) => BufReader::new(file),
            Err(e) => {
                error!("Failed to open registrations file: {e}");
                return HashMap::new();
            }
        };
```